keps/752-localqueue-weight/README.md

```diff
@@ -0,0 +1,309 @@
+tags, and then generate with `hack/update-toc.sh`.
+-->
+
+<!-- toc -->
+- [KEP-752: Support Weight in localQueue for better resource partitioning](#kep-752-support-weight-in-localqueue-for-better-resource-partitioning)
+  - [Summary](#summary)
+  - [Motivation](#motivation)
+    - [Goals](#goals)
+    - [Non-Goals](#non-goals)
+  - [Proposal](#proposal)
+    - [User Stories (Optional)](#user-stories-optional)
+      - [Story 1](#story-1)
+    - [Notes/Constraints/Caveats (Optional)](#notesconstraintscaveats-optional)
+    - [Risks and Mitigations](#risks-and-mitigations)
+  - [Design Details](#design-details)
+    - [API](#api)
+    - [Support Weight in localQueue](#support-weight-in-localqueue)
+    - [Local Queue Filtering and Sorting Policy Interface](#local-queue-filtering-and-sorting-policy-interface)
+    - [Extension Points in Cluster Queue](#extension-points-in-cluster-queue)
+    - [Test Plan](#test-plan)
+        - [Prerequisite testing updates](#prerequisite-testing-updates)
+      - [Unit Tests](#unit-tests)
+      - [Integration tests](#integration-tests)
+    - [Graduation Criteria](#graduation-criteria)
+  - [Implementation History](#implementation-history)
+  - [Drawbacks](#drawbacks)
+  - [Alternatives](#alternatives)
+<!-- /toc -->
+
+## Summary
+
+<!--
+This section is incredibly important for producing high-quality, user-focused
+documentation such as release notes or a development roadmap. It should be
+possible to collect this information before implementation begins, in order to
+avoid requiring implementors to split their attention between writing release
+notes and implementing the feature itself. KEP editors and SIG Docs
+should help to ensure that the tone and content of the `Summary` section is
+useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+
+Both in this section and below, follow the guidelines of the [documentation
+style guide]. In particular, wrap lines to a reasonable length, to make it
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+Add a extension point to determine whether a workload in certain local queue can
+be scheduled.  
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+Currently, for queues inside one CQ, we can submit queues as much as we want until the bottleneck of CQ, and one queue can occupy the whole resources with higher priorities. A limitation for a local queue is helpful for resource partitioning.
```

Comment:
IMHO, there are two options here:

* stay with 2 level hierarchy that Kueue offers today - have each local queue point to a different cluster queue with some resources, and allow some borrowing/eviction between queues via cohorts.
* explicitly introduce multi-level hierarchy

This KEP is somewhere between the two - introduces 3 level hierarchy, but with a different functionality set on level 1 (local queues in cluster queue) and level 2 (cluster queues in a cohort).

Reply:
To Marcin's point, one thing to consider is that LocalQueues are namespaced. As such, it means that administrators could only define weights between users that are in the same namespace.

This could be quite limiting for some organizations that requires a namespace for each end-user.

Reply:
Sorry maybe I didn't totally understand u. Do you mean maybe we should consider adding weight on cluster queues instead of local queues?

Reply:
I think weight in local queue is a resource consumption policy to control whose priority is more higher when scheduling and the resources in the queue will not be further segmented. We will still stay with 2 level hierarchy that Kueue offers today.

Reply:
Today, fair sharing and borrowing semantics happen for CQs within a cohort.

By adding weight to LQs, we are adding fair sharing to that level, implicitly adding one more level of hierarchy. The question is: do we want to do that, or should we rather implement arbitrary hierarchies where CQs can belong to other CQs or Cohort is an CRD and cohorts can belong to cohorts.

I'm not saying that one option or the other is better. I'm saying that we need to ponder all the options. Kueue APIs offer backwards compatibility, meaning that if we add an API, we need to maintain it forever or offer a deprecation process.

> To Marcin's point, one thing to consider is that LocalQueues are namespaced. As such, it means that administrators could only define weights between users that are in the same namespace.
> This could be quite limiting for some organizations that requires a namespace for each end-user.

Nvm about that comment. I realize it's wrong as multiple namespaces can point to the same CQ and we could respect weight among them.

Reply:
The goal of local queues is to offer batch users a way to fine tune how they want share the resources between their workloads without impacting the guarantees the admin gave to other tenants in other namespaces. 

My initial assumption is that local queues can be created and modified by the user (e.g., create the local queue to manage an experiment that includes running a sequence of jobs, but they want to limit how much resources the experiment should get), and so the admin ideally shouldn't need to care too much about them. So I think we need to offer quotas at the local queue level, but that is a knob for the batch user to fine tune resource sharing between their workloads, not for the admin to tune sharing between tenants.

The proposal in this KEP breaks the above resource model, and I don't think that is a good thing.

Reply:
Thanks for explain! 
In my cloud scene, my customers will create a quota node for each department and create a leaf quota node for each user, each user will have their own namespace to submit workloads. Unused resources will be shared by users by min and max like elastic quota in scheduler-plugins.
So I think maybe a 3 level tree is enough for the cases. 

I have three ideas.
1. Add a crd to define weights for local queues in a certain namespace and allow users to create and modify the resource may meet the requirement of Abdullah. But will still add one more level of hierarchy.

2. Allow admin to define parent cluster queue in cluster queue's spec. Cohort can only be set for the cluster queues who have same parent. Sum nominal resources of Children CQ must less than parent CQ's nominal resources. In this case, the functions in this KEP can be converted into multi-level CQ. But because users can not create and modify cluster queue, users will not be able to manage different workloads' resource consumption of their own.

3. Combine 1 and 2. 

HDYT?

Reply:
@alculquicondor @ahg-g Hi, if u have time, please take a look and let me know your options, I will continue work on this issue, thank very much. :D

Reply:
I think the better approach would be something along the lines of 2, which is also similar to #79

And here we have two high level alternatives:
a. Make a cohort a CRD. Then you could make cohorts be part of other cohorts. In this case, cohorts themselves don't have quota (like today). But they could still have weights, that only have semantics around borrowing.
b. Make CQs either belong to a cohort (today's mode) or another CQ. In this case, the child CQ cannot have limits, but can only have weights.

It might be useful to look into how systems like YARN or Slurm do multi-level hierarchy.

Question about your use-case: why not create a cohort per department and one CQ per user?

Reply:
Also #62 is related.

Reply:
Ok, thanks, if there is any progress, I will add it on KEP